code/controllers/subsystem/movement/movement.dm

```diff
@@ -22,30 +33,75 @@ SUBSYSTEM_DEF(movement)
 
 	var/datum/controller/subsystem/movement/old_version = global.vars["SS[our_name]"]
-	processing = old_version.processing
-	currentrun = old_version.currentrun
+	buckets = old_version.buckets
+	sorted_buckets = old_version.sorted_buckets
 
 /datum/controller/subsystem/movement/fire(resumed)
 	if(!resumed)
 		canonical_time = world.time
-		currentrun = processing.Copy()
-
-	var/list/running = currentrun //Cache for... you've heard this before
-	while(running.len)
-		var/datum/move_loop/loop = running[running.len]
-		running.len--
-		if(loop.timer <= canonical_time)
-			loop.process() //This shouldn't get nulls, if it does, runtime
-		if (MC_TICK_CHECK)
+
+	for(var/list/bucket_info as anything in sorted_buckets)
+		var/time = bucket_info[MOVEMENT_BUCKET_TIME]
+		if(time > canonical_time || MC_TICK_CHECK)
 			return
+		pour_bucket(bucket_info)
+
+/// Processes a bucket of movement loops (This should only ever be called by fire(), it exists to prevent runtime fuckery)
+/datum/controller/subsystem/movement/proc/pour_bucket(list/bucket_info)
+	var/list/processing = bucket_info[MOVEMENT_BUCKET_LIST] // Cache for lookup speed
+	while(processing.len)
+		var/datum/move_loop/loop = processing[processing.len]
+		processing.len--
+		loop.process() //This shouldn't get nulls, if it does, runtime
+		if(!QDELETED(loop)) //Re-Insert the loop
+			loop.timer = world.time + loop.delay
+			queue_loop(loop)
+		if (MC_TICK_CHECK)
+			break
+
+	if(length(processing))
+		return // Still work to be done
+	var/bucket_time = bucket_info[MOVEMENT_BUCKET_TIME]
+	smash_bucket(1, bucket_time) // We assume we're the first bucket in the queue right now
 	visual_delay = MC_AVERAGE_FAST(visual_delay, max((world.time - canonical_time) / wait, 1))
 
+/// Removes a bucket from our system. You only need to pass in the time, but if you pass in the index of the list you save us some work
+/datum/controller/subsystem/movement/proc/smash_bucket(index, bucket_time)
+	if(!index)
+		for(var/i in 1 to length(sorted_buckets))
+			var/list/bucket_info = sorted_buckets[i]
+			if(bucket_info[MOVEMENT_BUCKET_TIME] != bucket_time)
+				continue
+			index = i
+			break
+	sorted_buckets.Cut(index, index + 1) //Removes just this list
+	//Removes the assoc lookup too
+	buckets -= "[bucket_time]"
+
+/datum/controller/subsystem/movement/proc/queue_loop(datum/move_loop/loop)
+	var/target_time = loop.timer
+	var/string_time = "[target_time]"
+	if(buckets[string_time])
+		buckets[string_time] += loop
+	else
+		buckets[string_time] = list(loop)
+		// This makes buckets and sorted buckets point to the same place, allowing for quicker inserts
+		var/list/new_bucket = list(list(target_time, buckets[string_time]))
+		BINARY_INSERT_DEFINE(new_bucket, sorted_buckets, SORT_VAR_NO_TYPE, list(target_time), SORT_FIRST_INDEX, COMPARE_KEY)
+
+/datum/controller/subsystem/movement/proc/dequeue_loop(datum/move_loop/loop)
+	var/list/our_entries = buckets["[loop.timer]"]
+	our_entries -= loop
+	if(!our_entries)
```

Comment:
I think this is wrong. Did you mean to check `!our_entries.len`? If `our_entries` could ever be null at this point, it would also throw a runtime error on the previous line when you try to remove something from it.

Reply:
This is wrong, you're right. In a perfect world I'd use the same logic in both places I smash buckets, but something something proc overhead. (Glad this was in the case that hardly anything actually uses though)

You wanna pr a fix for it?

Reply:
Just changing this to `!our_entries.len` results in runtimes like this one:

```
[13:31:06] Runtime in movement.dm, line 77: list index out of bounds
proc name: smash bucket (/datum/controller/subsystem/movement/proc/smash_bucket)
src: Space Drift (/datum/controller/subsystem/movement/spacedrift)
call stack:
Space Drift (/datum/controller/subsystem/movement/spacedrift): smash bucket(1, 267.5)
Space Drift (/datum/controller/subsystem/movement/spacedrift): pour bucket(/list (/list))
Space Drift (/datum/controller/subsystem/movement/spacedrift): fire(0)
Space Drift (/datum/controller/subsystem/movement/spacedrift): ignite(0)
Master (/datum/controller/master): RunQueue()
Master (/datum/controller/master): Loop()
Master (/datum/controller/master): StartProcessing(0)
```

I don't understand this system at all, so I don't think I can be of much more help.

Reply:
MMM, that's worrying.
Shouldn't be too much of a headache to deal with, but if you're not comfortable I can handle it
Thanks a ton for the report

Reply:
I investigated this during my port to bee and there are some issues where the two lists start to get asyncron so an item that should be deleted does not exist in the other list anymore the exact reason for that i am not sure my solution was to merge the functions of both lists into a single one and use that one instead.

Reply:
I think that might make lookup slower? not totally sure why I avoided it.
Should really get around to fixing it properly

Reply:
the only real concern for slowdown i have is during the list sort because i basically have to convert the assoc value of the older list entries into a number to make the comparison

Reply:
Oh I see it involves a text2num, that's a bit unfortunate, but not a huge issue. It's weird to me that you could end up in a position where you're smashing a bucket that doesn't exist...